doc/parameter_server.md

```diff
@@ -0,0 +1,144 @@
+- pserver, maintaining the global model, and providing parameter pull/push/optimize/checkpoint service
+- worker, computing gradients of model parameters using local data and pushing gradients to pservers
+
+
+Please refer to:
+
+![parameter_server](./images/parameter_server.png)
+
+## Key concepts
+
+
+### Parameter sharding
+
+
+We prefer to shard the global model for some reasons:
+
+- A model could be too large to fit in the memory of a process. For example, many recommending and ranking models take very high-dimensional (and sparse) inputs, thus require large embedding tables. In such cases, we'd have to have multiple parameter server instances to hold the global model.
+
+- Even when a single process can hold the global model, we might still shard it onto multiple parameter server instances, who share the both communication and optimization workload.
+
+There are several kinds of parameter to be handled separately:
+
+- Very big embedding table: Embedding table is a collection of <item id, embedding vector> pairs. There is also a mapping from item id to pserver id, so the corresonding embedding vector will be stored at the certain pserver
+
+- Big dense tensor: If the dense tensor parameter exceeds certain size, it will be sliced into several subtensors. The parameter name combining subtensor id will become a unique key, and the value is the subtensor(Please note that the slice tensor operation is zero-copy)
+
+- Small dense tensor: The small dense tensor parameter will be stored at certain pserver wholely
+
+
+The local model on workers and the global model on the parameter server(s) have the same size. In case that large embedding tables make the model size too big to fit in a single process's memory space, an intuitive solution is to save the model copies on an external storage service like Redis or Memcached.
+
+However, such a solution creates a complex workflow which introduces too many cross-node communications.
+
+We propose an alternative solution that doesn't rely on Redis or Memcachd. Instead, the global model, including the large embedding tables, is sharded across parameter servers, and the local model contains only part of the embedding table -- the small fraction that is required to compute recent minibatch(es).
+
+### Parameter Initialization
+
+
+Parameter initialization of a very big embedding table is lazy. For example, in online learning, there could be unkown item id in the training data. So, until worker send the unkown item id to pserver, will pserver initialize corresponding embedding vector and send back to worker. This is a `get_or_create` semantic.
+
+Other parameters could be initialized before training.
+
+
+## Workflow
+
+### Assumption
+
+We apply different solutions to different scenarios.
+
+
+| scenario| hardware | update stragety | communication  stragety |
+| :----  |:----  |:---|:---|
+| Recommendation  | CPU|  Async |  Parameter server |
+| CV | GPU | BSP| Allreduce|
```

Comment:
Need an explanation of what BSP is short for.

Reply:
Done